wingtips-core/src/test/java/com/nike/wingtips/TraceAndSpanIdGeneratorTest.java

```diff
@@ -143,4 +187,68 @@ public void generateId_should_not_generate_duplicate_ids_over_reasonable_number_
         // then
         assertThat(ids.size()).isEqualTo(numAttempts);
     }
+
+    @Test
+    public void generate64BitRandomLong_should_not_generate_duplicate_ids_over_reasonable_number_of_attempts() throws Exception {
+        // given
+        Set<Long> randomLongs = new HashSet<>();
+        int numAttempts = 1000000;
+
+        // when
+        for (int i = 0; i < numAttempts; i++) {
+            randomLongs.add(TraceAndSpanIdGenerator.generate64BitRandomLong());
+        }
+
+        // then
+        assertThat(randomLongs.size()).isEqualTo(numAttempts);
+    }
+
+    @DataProvider(value = {
+        "0000000000000000   |   0",
+        "0000000000000001   |   1",
+        "ffffffffffffffff   |   18446744073709551615",
+        "fffffffffffffffe   |   18446744073709551614",
+        "7fae59489091369a   |   9200389256962455194",
+        "eb5e7aaefeb92b4f   |   16960128138740312911",
+        "d2153abe4c047408   |   15138070311469347848",
+        "9041ee0d07d6c72c   |   10394851154681317164",
+        "6470a5ce0e9262f4   |   7237466905610707700",
+        "000003c8a251fb93   |   4160251624339",
+    }, splitBy = "\\|")
+    @Test
+    public void longToUnsignedLowerHexString_and_unsignedLowerHexStringToLong_work_as_expected_for_known_values(String actualHexValue, String actualUnsignedDecimalValue) {
+        // given
+        long actualSignedPrimitive = new BigInteger(actualUnsignedDecimalValue).longValue();
+
+        // when
+        String calculatedHexValue = TraceAndSpanIdGenerator.longToUnsignedLowerHexString(actualSignedPrimitive);
+        long calculatedPrimitiveValue = TraceAndSpanIdGenerator.unsignedLowerHexStringToLong(actualHexValue);
+
+        // then
+        assertThat(calculatedHexValue).isEqualTo(actualHexValue);
+        assertThat(calculatedPrimitiveValue).isEqualTo(actualSignedPrimitive);
+    }
+
+    @DataProvider(value = {
+        "                   ", // less than 16 chars
+        "12345678901234567  ", // longer than 16 chars
+        "/                  ", // before '0' char
+        ":                  ", // after '9' char
+        "`                  ", // before 'a' char
+        "g                  ", // after 'f' char
+        "ABCDEF             "  // uppercase hex chars
+    }, splitBy = "\\|")
+    @Test
+    public void unsignedLowerHexStringToLong_throws_NumberFormatException_for_illegal_args(final String badHexString) {
+        // when
+        Throwable ex = catchThrowable(new ThrowableAssert.ThrowingCallable() {
```

Comment:
retrolamba is pretty cool. it rewrites your bytecode back to 6 or 7! This
would eliminate the need to have gradle know anything clever.

On Sat, Aug 20, 2016 at 9:47 AM, Nic Munroe <notifications@github.com>
wrote:

> In wingtips-core/src/test/java/com/nike/wingtips/
> TraceAndSpanIdGeneratorTest.java
> <https://github.com/Nike-Inc/wingtips/pull/16#discussion_r75570424>:
>
> > +        assertThat(calculatedPrimitiveValue).isEqualTo(actualSignedPrimitive);
> > +    }
> > +
> > +    @DataProvider(value = {
> > +        "                   ", // less than 16 chars
> > +        "12345678901234567  ", // longer than 16 chars
> > +        "/                  ", // before '0' char
> > +        ":                  ", // after '9' char
> > +        "`                  ", // before 'a' char
> > +        "g                  ", // after 'f' char
> > +        "ABCDEF             "  // uppercase hex chars
> > +    }, splitBy = "\\|")
> > +    @Test
> > +    public void unsignedLowerHexStringToLong_throws_NumberFormatException_for_illegal_args(final String badHexString) {
> > +        // when
> > +        Throwable ex = catchThrowable(new ThrowableAssert.ThrowingCallable() {
>
> Yeah I like the idea. I may start doing that in the future when feel like
> I have the time to wrangle gradle to make the tests work in java 8 but
> guarantee the libraries still build in java 7.
>
> Really I'd rather just move the whole thing to java 8. =p Once things are
> stable I may do that and just have a legacy java 7 version that is mostly
> stagnant for people stuck on stacks that can't go to java 8.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Nike-Inc/wingtips/pull/16/files/f037fd2cb5d00e196447cbaa83a39de830394e7b#r75570424>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAD618z-EzQoikIusfsyfxITLmjx9jFaks5qhlylgaJpZM4Jo1BS>
> .
>